doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
 
-The reverse operation (writing data to a Ptr{T}), can be performed using
-``unsafe_store!(ptr, value, [index])``.  Currently, this is only supported
-for bitstypes or other pointer-free (``isbits``) immutable types.
+Struct Type correspondences
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Any operation that throws an error is probably currently unimplemented
-and should be posted as a bug so that it can be resolved.
+Composite types, aka ``struct`` in C or ``STRUCTURE`` / ``RECORD`` in Fortran),
+can be mirrored in Julia by creating a ``type`` or ``immutable``
+definition with the same field layout.
 
-If the pointer of interest is a plain-data array (bitstype or immutable), the
-function ``pointer_to_array(ptr,dims,[own])`` may be more useful. The final
-parameter should be true if Julia should "take ownership" of the underlying
-buffer and call ``free(ptr)`` when the returned ``Array`` object is finalized.
-If the ``own`` parameter is omitted or false, the caller must ensure the
-buffer remains in existence until all access is complete.
+When used recursively, ``isbits`` types are stored inline.
+All other types are stored as a pointer to the data.
+When mirroring a struct used by-value inside another struct in C,
+it is imperative that you do not attempt to manually copy the fields over,
+as this will not preserve the correct field alignment.
+Instead, declare an immutable isbits type and use that instead.
+Unnamed structs are not possible in the translation to Julia.
 
-Arithmetic on the ``Ptr`` type in Julia (e.g. using ``+``) does not behave the
-same as C's pointer arithmetic. Adding an integer to a ``Ptr`` in Julia always
-moves the pointer by some number of *bytes*, not elements. This way, the
-address values obtained from pointer arithmetic do not depend on the
-element types of pointers.
+Packed structs and union declarations are not supported by Julia.
+
+You can get a near approximation of a ``union`` if you know, a priori,
+the field that will have the greatest size (potentially including padding).
+When translating your fields to Julia, declare the Julia field to be only
+of that type.
+
+Arrays of parameters must be expanded manually, currently
+(either inline, or in an immutable helper-type). For example::
+
+    in C:
+    struct B {
+        int A[3];
+    };
+    b_a_2 = B.A[2];
+
+    in Julia:
+    immutable B_A
+        A_1::Cint
+        A_2::Cint
+        A_3::Cint
+    end
+    type B
+        A::B_A
+    end
+    b_a_2 = B.A.(2)
+
+Arrays of unknown size are not supported.
+
+In the future, some of these restrictions may be reduced or eliminated.
+
+Memory Ownership:
+~~~~~~~~~~~~~~~~~
+
+**malloc/free**
+
+Memory allocation and deallocation of such objects must be
+handled by calls to the appropriate cleanup routines in the libraries
+being used, just like in any C program. Do not try to free an object
+received from a C library with `c_free` in Julia, as this may result
```

Comment:
Correct. Technically that is a bug in their library (Unless that library happened to be called libc...)